# Laboratorio de regresión - 3

## Significancia de factores

|                |   |
:----------------|---|
| **Nombre**     | Julieta Madrigal Florea  |
| **Fecha**      |  Lunes 2 de febrero 2026 |
| **Expediente** | 744029  |

Descarga el archivo de publicidad y carga los datos (Advertising.csv).

In [1]:
import pandas as pd
import numpy as np

In [2]:
data = pd.read_csv('Advertising.csv')
data.head()

Unnamed: 0     TV  radio  newspaper  sales
0           1  230.1   37.8       69.2   22.1
1           2   44.5   39.3       45.1   10.4
2           3   17.2   45.9       69.3    9.3
3           4  151.5   41.3       58.5   18.5
4           5  180.8   10.8       58.4   12.9

**¿Hay alguna relación entre el presupuesto para publicidad y las ventas?**

Nuestra primera meta debe ser determinar si hay evidencia en los datos de que haya una asociación entre estas variables.

- ¿Por qué? ¿Qué resultaría si nos diéramos cuenta de la falta de relación entre el presupuesto de publicidad y las ventas?

**R =** Si nos diéramos cuenta de la falta de relación entre el presupuesto de publicidad y las ventas nos daríamos cuenta de que no vale la pena seguir invirtiendo en eso pero a largo plazo, esto permite crear una estrategia para encontrar lo que realmente le funciona al negocio.

**¿Qué tan fuerte es esta relación?**
Asumiendo que existe esta relación, ¿nos sirve conocer el impacto que tiene invertir en publicidad en las ventas?

**R=** Conocer el impacto que tiene invertir en publicidad en las ventas permite saber cuánto se debe de invertir en cada medio para alcanzar la meta de ventas; además, se podría calcular cuánta ganacia se genera por unidad invertida en publicidad e incluso hacer una proyección basandose en el presupuesto de publicidad.

**¿Cuáles medios están asociados con las ventas? ¿Qué tan grande es la asociación entre un medio específico y las ventas?**

Hay 3 medios distintos en los datos. ¿Sirve invertir en los 3? ¿Conviene más invertir sólo en uno?

**¿Qué tan seguros estamos de que podríamos predecir ventas futuras?**

**¿La relación es lineal?**

**¿Hay sinergia entre estos medios?**

Puede ser que gastar \\$50,000 en publicidad y otros \\$50,000 en radio es mejor opción que gastar \\$100,000 en televisión. A esto le llamamos efecto de interacción.

Podemos usar regresión lineal para responder todas estas preguntas.

Realiza una regresión lineal:

$$ \text{ventas} \approx \beta_0 + (\beta_1)(\text{TV})$$

In [3]:
# VENTAS
y = data['sales']
y = np.array(y).reshape(-1, 1)

In [4]:
# Factor TV
x = data['TV']
x = np.array(x).reshape(-1, 1)

In [5]:
from sklearn.linear_model import LinearRegression

In [6]:
lr = LinearRegression()
lr.fit(x, y)

LinearRegression()

In [7]:
# beta 0 = lo que vendo sin publicidad
beta_0 = lr.intercept_
beta_0

array([7.03259355])

$\hat{\beta_0}$ = 7.0326

In [8]:
# beta 1 = TV
# por cada unidad de TV, se va a ganar 0.04 mas
beta_1 = lr.coef_
beta_1

array([[0.04753664]])

$\hat{\beta_1}$ = 0.0475

### Verificando la precisión de nuestros coeficientes estimados

Recuerda que en el mundo real hay ruidos y errores de medición. Siempre se asume que la verdadera relación entre $X$ y $Y$ es $$Y = \beta_0 + \beta_1 X + \epsilon$$

Se asume que el término de error es independiente de $X$ (el error siempre es el mismo sin importar el valor de $X$). Este modelo describe a la *línea de regresión de la población*, que es la mejor aproximación de la verdadera relación entre $X$ y $Y$. Cuando usamos mínimos cuadrados encontramos la *línea de mínimos cuadrados*.

¿Cuál es la diferencia entre población y muestra?

**R =** Población es el conjunto completo de todos los elemnetos que se quieren estudiar; es decir, se incluyen todos los datos del grupo de interés. Mientras que la muestra es una parte o subconjunto de la población, solo incluye a ciertos elementos de la población.

¿Cuál crees que sea la diferencia entre hacer una regresión con todos los datos de la población y una muestra de ella?

**R =** La diferencia entre hacer regresión con todos los datos de la población y una muestra de ella es que si se utiliza la población, los resultados obtenidos serían exactos y reales porque no habría margen de error; además, no se necesitarían intervalos de confianza ni pruebas de hipotésis porque ya tengo toda la información y por ende, los resultados no son estimaciones, son definitivos. Por otro lado, si se utiliza una muestra los resultados son aproximaciones porque se generalizan los resultados de toda la población, es por eso que los intervalos de confianza y pruebas de hipotésis determinan si las relaciones son significativas.

La línea de regresión de la población no se puede observar. El concepto de comparar estas líneas es una extensión natural del acercamiento estadístico estándar de usar información de una muestra para estimar características de una población grande.

Imagina que quieres encontrar la altura promedio de un mexicano $\mu$. Medir a todos y cada uno de los mexicanos en situaciones similares, con la misma regla, mismo operador, y otras incontables formas de minimizar la variación de la medida es una tarea imposible. Lo que podemos asumir es que $\hat{\mu} = \bar{y}$. La media poblacional y la media muestral son diferentes, pero la media muestral es usualmente un buen estimado.

De la misma manera, como no contamos con el 100% de la información para hacer una regresión, los coeficientes $\beta_0$ y $\beta_1$ son desconocidos. Podemos estimarlos usando mínimos cuadrados, encontrando $\hat{\beta_0}$ y $\hat{\beta_1}$. Puede que las muestras que tengamos en ese momento estén un poco por encima de la media, pero otras muestras en otro momento puede que estén debajo de la media. En general, esperamos que el promedio de las aproximaciones $\hat{\mu}$ aproxime a $\mu$.

Esto lleva a la pregunta: ¿qué tan cercanos son nuestros coeficientes estimados a los verdaderos coeficientes? Utilizamos el concepto de error estándar para evaluar esto.

$$ \text{Var}(\hat{\mu})=\text{SE}(\hat{\mu})^2 = \frac{\sigma^2}{n} $$

Donde $\sigma$ es la desviación estándar de cada una de las observaciones $y_i$ de $Y$. El error estándar nos dice la cantidad promedio que el estimado difiere del valor verdadero. Podemos ver en la fórmula que entre más observaciones tengamos el error se hace más pequeño. Las fórmulas para errores estándar de $\hat{\beta_0}$ y $\hat{\beta_1}$ son:

$$ \text{SE}(\hat{\beta_0})^2 = \sigma^2 [\frac{1}{n} + \frac{\bar{x}^2}{\sum_{i=1}^n (x_i - \bar{x})^2}]$$

$$ \text{SE}(\hat{\beta_1})^2 = \frac{\sigma^2}{\sum_{i=1}^n (x_i - \bar{x})^2}$$

$$ \sigma^2 = \text{Var}(\epsilon) = \text{RSE}^2 = \frac{\text{RSS}}{n-p}$$

Para que estas fórmulas sean validas asumimos que los errores $\epsilon_i$ tienen varianza común $\sigma^2$ y que no están correlacionados.

Calcula los errores estándar de los coeficientes

In [9]:
# n es el tamaño de la muestra
# p
n,p = x.shape

Calculamos **$\bar{x}$**

In [10]:
x_bar = np.mean(x)

Calculamos **$RSS = \sum (y - \hat{y})^2$**

In [11]:
y_pred = lr.predict(x)

In [12]:
rss = sum((y - y_pred)**2)
rss

array([2102.53058313])

Entonces **$ \sigma^2 = \text{Var}(\epsilon) = \text{RSE}^2 = \frac{\text{RSS}}{n-p}$**

In [13]:
var = rss / (n - p)
var

array([10.56548032])

Para calcular:

$$\text{SE}(\hat{\beta_0})^2 = \sigma^2 \left[\frac{1}{n} + \frac{\bar{x}^2}{\sum_{i=1}^n (x_i - \bar{x})^2}\right] \qquad \text{y} \qquad \text{SE}(\hat{\beta_1})^2 = \frac{\sigma^2}{\sum_{i=1}^n (x_i - \bar{x})^2}$$

Primero calculamos **${\sum_{i=1}^n (x - \bar{x})^2}$** y le vamos a llamar "aux"

In [14]:
aux = np.sum((x - x_bar)**2)

Ahora si podemos calcular $\text{SE}(\hat{\beta_0})$

In [15]:
# le quitamos el cuadrado con una raíz cuadrada para simplificar
se_b0 = np.sqrt(var * ((1/n) + ((x_bar)**2 / aux )))
se_b0

array([0.45669113])

Se hace lo mismo para $ \text{SE}(\hat{\beta_1})$

In [16]:
se_b1 = np.sqrt(var / aux)
se_b1

array([0.00268384])

Estos errores se pueden usar para calcular intervalos de confianza. Un intervalo de confianza del $95\%$ se define como un rango de valores en el cuál se encuentra el desconocido valor verdadero con un $95\%$ de probabilidad.

Otra forma de verlo es que si tomamos muestras repetidas y construimos un intervalo de confianza para cada una, el $95\%$ de los intervalos creados van a contener el valor verdadero. Para la regresión el intervalo de confianza del $95\%$ toma la forma:

$$ \hat{\beta_j} \pm 2\text{SE}(\hat{\beta_j})$$

Calcula los intervalos de confianza al **95%** para los coeficientes estimados:

Primero para $\text{SE}(\hat{\beta_0})$

In [17]:
beta_0PLUS = beta_0 + 2*se_b0
beta_0PLUS

array([7.94597581])

In [18]:
beta_0MINUS = beta_0 - 2*se_b0
beta_0MINUS

array([6.11921128])

El intervalo de confianza de $$\text{SE}(\hat{\beta_0}) = \qquad \text{[ 6.12} \qquad  7.95]$$

Ahora $ \text{SE}(\hat{\beta_1})$

In [19]:
beta_1PLUS = beta_1 + 2*se_b1
beta_1PLUS

array([[0.05290432]])

In [20]:
beta_1MINUS = beta_1 - 2*se_b1
beta_1MINUS

array([[0.04216896]])

El intervalo de confianza de $$\text{SE}(\hat{\beta_1}) = \qquad \text{[ 0.042} \qquad  0.053]$$

Los errores estándar también se usan para realizar pruebas de hipótesis. La prueba de hipótesis más común es probar la hipótesis nula de:

$$ H_0: \text{No hay relación entre } X \text{ y } Y \ \ \ \ (\beta_1=0)$$

contra la hipótesis alternativa:
$$ H_0: \text{Hay alguna relación entre } X \text{ y } Y \ \ \ (\beta_1 \neq 0)$$

Explica con tus palabras el significado de la hipótesis nula y la hipótesis alternativa.

**R =** Por ejemplo, si quiero demostrar que la TV tiene relación con las ventas, mi hipótesis nula es que NO existe relación y mi hipótesis alternativa es que SI existe relación; es decir, la hipótesis alterntiva es que lo que quiero demostrar SI existe, mientras que la hipótesis nula es que lo que quiero demostrar NO existe.

Para probal la hipótesis nula debemos determinar si nuestro estimado $\hat{\beta_1}$ de $\beta_1$ está lo suficientemente alejado de cero para que podamos decir con confianza que este valor no es cero.

¿Qué tan lejos? Depende de qué tanta confianza tengamos en el estimado encontrado. Si nuestro error estándar es pequeño y nuestro estimado está alejado de cero podríamos decir que hay muy poca probabilidad de que el valor verdadero sea 0. En cambio, si nuestro error estándar es grande y nuestro estimado está muy cerca de cero, entonces podrías ser que el valor verdadero sea cero y que no haya relación entre las variables.

Se calcula un *estadístico t* dado por
$$ t = \frac{\hat{\beta_j} - \mu}{\text{SE}(\hat{\beta_j})} $$

donde $\mu$ es el valor contra el que queremos probar.

Calcula el estadístico t para tus coeficientes estimados, usando como referencia la prueba de hipótesis.

Como se está probando si existe o no una relación **μ = 0** y mi hipótesis nula **H₀: β₁ = 0**, es decir no hay relación.

In [21]:
# si en la hipotésis nula ponemos otro número diferente de 0, mu sería = 2,3,4
# depende de contra que valor estamos comprando
mu = 0

Para $ \text{SE}(\hat{\beta_0})$

t = 15.4

In [22]:
t_b0 = (beta_0 - mu ) / se_b0
t_b0

array([15.3990149])

Para $ \text{SE}(\hat{\beta_1})$

t = 17.71

In [23]:
t_b1 = (beta_1 - mu ) / se_b1
t_b1

array([[17.71218463]])

La distribución t tiene forma de campana y se parece bastante a la distribución normal cuando $n > 30$. Ya sólo es cuestión de calcular la probabilidad de observar cualquier número tal que su valor absoluto sea igual o mayor que el valor absoluto del estadístico t calculado. En otras palabras:
$$ P(|x| \geq |t|) $$

A esta probabilidad la llamamos *p-value*. Un *p-value* pequeño indica que es poco probable que exista por puro azar una relación significativa entre predictor y respuesta, en caso de que no haya una asociación real entre predictor y respuesta. En otras palabras, el *p-value* te dice la probabilidad de que parezca que hay relación cuando no la hay.

Si el *p-value* es pequeño, inferimos que sí hay una asociación entre el predictor y la respuesta, y **rechazamos la hipótesis nula**.
  

¿Qué tan pequeño? Depende de la aplicación. Un valor muy común es del $5\%$.

Utiliza el siguiente código para calcular el *p-value* para tus coeficientes

`from scipy import stats`

`p_bj = 2*(1 - stats.t.cdf(np.abs(t_bj), n-p))`

In [24]:
from scipy import stats

P-value para $ \hat{\beta_0}$

In [25]:
p_b0 = 2*(1 - stats.t.cdf(np.abs(t_b0), n-2))
p_b0

array([0.])

P-value para $ \hat{\beta_1}$

In [26]:
p_b1 = 2*(1 - stats.t.cdf(np.abs(t_b1), n-2))
p_b1

array([[0.]])

¿Se rechaza la hipótesis nula? ¿Qué significa?

**R =** Se rechaza la hipotésis nula porque 0 es menor a 0.05, por lo tanto se acepta la hipotésis alterna y significa que la beta 0 real y la beta 1 real son diferentes de 0; es decir, si hay relación significativa entre la inversión en TV y las ventas. Hay 0% de probabilidad de que el resultado de beta 0 y beta 1 fuera casualidad.

Realiza otras dos regresiones. Ya tienes hecha la regresión de ventas dado el gasto en publicidad de TV. Realiza la regresión para gastos en radio y gastos en periódico. Organiza las respuestas para que debajo de esta celda se tenga:
- Título de regresión
- Coeficientes estimados
- Errores estándar de los coeficientes
- Intervalos de confianza
- Estadísticos t
- p-values
- Observaciones

## Radio

- Coeficientes estimados

In [27]:
r = data['radio']
r = np.array(r).reshape(-1, 1)

In [28]:
from sklearn.linear_model import LinearRegression

In [29]:
lr = LinearRegression()
lr.fit(r, y)

LinearRegression()

In [30]:
beta_0_r = lr.intercept_
beta_0_r

array([9.3116381])

In [31]:
beta_1_r = lr.coef_
beta_1_r

array([[0.20249578]])

- Errores estándar de los coeficientes

In [32]:
n_r,p_r = r.shape

In [33]:
r_bar = np.mean(r)

In [34]:
y_pred_r = lr.predict(r)

In [35]:
rss_r = sum((y - y_pred_r)**2)
rss_r

array([3618.47954903])

In [36]:
var_r = rss_r / (n_r - p_r)
var_r

array([18.18331432])

In [37]:
aux_r = np.sum((r - r_bar)**2)

In [38]:
se_b0_r = np.sqrt(var_r * ((1/n_r) + ((r_bar)**2 / aux_r )))
se_b0_r

array([0.56148439])

In [39]:
se_b1_r = np.sqrt(var_r / aux_r)
se_b1_r

array([0.02035996])

- Intervalo de confianza al 95%

In [40]:
beta_0_r + 2*se_b0_r

array([10.43460688])

In [41]:
beta_0_r - 2*se_b0_r

array([8.18866931])

El intervalo de beta 0 es **[8.19____10.43]**

In [42]:
beta_1_r + 2*se_b1_r

array([[0.2432157]])

In [43]:
beta_1_r - 2*se_b1_r

array([[0.16177587]])

El intervalo de beta 1 **[ 0.162___0.243 ]**

- Estadísticos t

In [44]:
# Es igual a 0 porque así se estableció en la hipotésis
mu

0

In [45]:
t_b0_r = (beta_0_r - mu ) / se_b0_r
t_b0_r

array([16.58396605])

In [46]:
t_b1_r = (beta_1_r - mu ) / se_b1
t_b1_r

array([[75.45006691]])

- P-value

In [47]:
# 2 porque solo es beta 0 y beta 1
factores = 2

In [48]:
p_b0_r = 2*(1 - stats.t.cdf(np.abs(t_b0_r), n_r-factores))
p_b0_r

array([0.])

In [49]:
p_b1_r = 2*(1 - stats.t.cdf(np.abs(t_b1_r), n_r-p))
p_b1_r

array([[0.]])

- Observaciones

El p-value de beta 0 y beta 1 es igual a 0, esto significa que tenemos que rechazar la hipotésis nula y aceptamos la hipotésis alterna porque mi beta 0 y beta 1 real NO son iguales a 0.

## Newspaper

In [50]:
news = data['newspaper']
news = np.array(news).reshape(-1, 1)

In [51]:
from sklearn.linear_model import LinearRegression

In [52]:
lr = LinearRegression()
lr.fit(news, y)

LinearRegression()

In [53]:
beta_0_news = lr.intercept_
beta_0_news

array([12.35140707])

In [54]:
beta_1_news = lr.coef_
beta_1_news

array([[0.0546931]])

- Errores estándar de los coeficientes

In [55]:
n_news,p_news = news.shape

In [56]:
news_bar = np.mean(news)

In [57]:
y_pred_news = lr.predict(news)

In [58]:
rss_news = sum((y - y_pred_news)**2)
rss_news

array([5134.80454411])

In [59]:
var_news = rss_news / (n_news - p_news)
var_news

array([25.80303791])

In [93]:
aux_news = np.sum((news - news_bar)**2)

In [61]:
se_b0_news = np.sqrt(var_news * ((1/n_news) + ((news_bar)**2 / aux_news )))
se_b0_news

array([0.61985686])

In [62]:
se_b1_news = np.sqrt(var_news / aux_news)
se_b1_news

array([0.01653402])

- Intervalo de confianza al 95%

In [63]:
beta_0_news + 2*se_b0_news

array([13.5911208])

In [64]:
beta_0_news - 2*se_b0_news

array([11.11169334])

El intervalo de beta 0 es **[11.11____13.6]**

In [65]:
beta_1_news + 2*se_b1_news

array([[0.08776114]])

In [66]:
beta_1_news - 2*se_b1_news

array([[0.02162505]])

El intervalo de beta 1 **[ 0.022___0.087]**

- Estadísticos t

In [67]:
# Es igual a 0 porque así se estableció en la hipotésis
mu

0

In [68]:
t_b0_news = (beta_0_news - mu ) / se_b0_news
t_b0_news

array([19.92622457])

In [69]:
t_b1_news = (beta_1_news - mu ) / se_b1_news
t_b1_news

array([[3.30791255]])

- P-value

In [70]:
# osea solo tienes beta 0 y beta 1
factores = 2

In [71]:
p_b0_news = 2*(1 - stats.t.cdf(np.abs(t_b0_news), n-factores))
p_b0_news

array([0.])

In [72]:
p_b1_news = 2*(1 - stats.t.cdf(np.abs(t_b1_news), n-factores))
p_b1_news

array([[0.00111636]])

- Observaciones

El p-value de beta 0 y beta 1 es igual a 0, esto significa que tenemos que rechazar la hipotésis nula y aceptamos la hipotésis alterna porque mi beta 0 y beta 1 real NO son iguales a 0. A pesar de que el p-value de mi beta 1 no es exactamente 0, el valor es muy cercano y se puede interpretar como p-value = 0.

## Regresión lineal múltiple

En lugar de hacer una regresión para cada factor independiente, quizás se puede extender el modelo para que tenga varios factores dentro:

$$ Y = \beta_0 + \beta_1 X_1 + \beta_2 X_2 + ... + \beta_p X_p + \epsilon $$

Para nuestro ejemplo de publicidad:

$$ \text{sales} = \beta_0 + \beta_1 (\text{TV}) + \beta_2 (\text{radio}) + \beta_3 (\text{newspaper}) + \epsilon $$

Utiliza la librería `statsmodels` para realizar la regresión. Por defecto la librería statsmodels no toma en cuenta el intercepto ($\beta_0$), por lo que se tendrá que agregar una columna de unos de tamaño *n* a la matriz X.

`import statsmodels.api as sm`

`ols = sm.OLS(Y, X)`

`results = ols.fit()`

`results.summary()`

In [73]:
import statsmodels.api as sm

In [74]:
# Hacemos la matriz con TV, radio y newspaper
Xcompleta = np.hstack((x, r, news))

In [75]:
# agregamos la columna de unos
Xcon_unos = np.column_stack([np.ones(len(Xcompleta)), Xcompleta])

In [76]:
ols = sm.OLS(y, Xcon_unos)

In [77]:
results = ols.fit()

In [78]:
# si dice uncentered = significa que no tiene la columna de unos
results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                      y   R-squared:                       0.897
Model:                            OLS   Adj. R-squared:                  0.896
Method:                 Least Squares   F-statistic:                     570.3
Date:                Sat, 31 Jan 2026   Prob (F-statistic):           1.58e-96
Time:                        03:35:50   Log-Likelihood:                -386.18
No. Observations:                 200   AIC:                             780.4
Df Residuals:                     196   BIC:                             793.6
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const          2.9389      0.312      9.422      0.000       2.324       3.554
x1             0.0458      0.001     32.809      0.000       0.043       0.049
x2             0.1885      0.009     21.893      0.000       0.172       0.206
x3            -0.0010      0.006     -0.177      0.860      -0.013       0.011
==============================================================================
Omnibus:                       60.414   Durbin-Watson:                   2.084
Prob(Omnibus):                  0.000   Jarque-Bera (JB):              151.241
Skew:                          -1.327   Prob(JB):                     1.44e-33
Kurtosis:                       6.332   Cond. No.                         454.
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

$$ \text{sales} = \text{2.9389} + \text{0.0458} (\text{TV }) + \text{0.1885} (\text{radio}) - \text{0.0010} (\text{newspaper}) + \epsilon $$

¿Qué diferencias puedes observar entre los *p-values* de una regresión múltiple y los encontrados en las regresiones simples? ¿Por qué crees que existen estas diferencias?

**TV (INDIVIDUAL)**

In [79]:
p_b0, p_b1

(array([0.]), array([[0.]]))

Al observar el p-value individual y los de una regresión múltiple, se puede observar que en ambas regresiones, el p-value es igual a 0, esto significa que SI existe relación entre TV y las ventas y existe suficiente información para considerar que la observación no fue al azar.

**RADIO (INDIVIDUAL)**

In [80]:
p_b0_r , p_b1_r

(array([0.]), array([[0.]]))

Al observar el p-value individual y los de una regresión múltiple, se puede observar que en ambas regresiones, el p-value es igual a 0, esto significa que SI existe relación entre la radio  y las ventas y existe suficiente información para considerar que la observación no fue al azar.

**NEWSPAPER (INDIVIDUAL)**

In [81]:
p_b0_news , p_b1_news

(array([0.]), array([[0.00111636]]))

Al observar el p-value individual y los de una regresión múltiple, se puede observar que en la regresión individual, newspaper SI tenía relación con las ventas

Sin embargo, al hacer la regresión múltiple se observa que newspaper tiene un p-value de 0.86, esto significa que existe un 86% de probabilidad de que las observaciones obtenidas hayan sido al azar y que su efecto o relevancia ya está capturado en radio y TV; es decir, ya no aporta información significativa / extra al modelo cuando existen más variables.

A partir de esto se puede concluir que newspaper no tiene un efecto definido cuando radio y TV se encuentran en la ecuación, de tal modo que si SOLO aumento la inversión a newspaper y no cambio TV ni radio, no se espera que las ventas aumenten.

## Referencia

James, G., Witten, D., Hastie, T., Tibshirani, R.,, Taylor, J. (2023). An Introduction to Statistical Learning with Applications in Python. Cham: Springer. ISBN: 978-3-031-38746-3

## Addendum

Para calcular los *p-values* de los parámetros sin `statsmodels`:
1. Calcular RSS
2. Calcular RSE
3. `var_beta = np.linalg.inv(X.T @ X) * rse**2` (X con columna de intercepto)
4. `std_beta = np.sqrt(var_beta.diagonal())` El orden de los valores corresponde al orden de los factores en las columnas de la matriz $X$.
5. Calcular *estadístico t*
6. Calcular *p-value*

In [82]:
# Matriz con unos, TV, radio y newspaper
Xcon_unos = np.column_stack([np.ones(len(Xcompleta)), Xcompleta])

In [83]:
n_multi,p_multi = Xcon_unos.shape

In [84]:
Xmulti_bar = Xcon_unos.mean()

In [87]:
from sklearn.linear_model import LinearRegression
lr_multi = LinearRegression()
lr_multi.fit(Xcon_unos, y)

LinearRegression()

In [115]:
beta_0_multi = lr_multi.intercept_
beta_0_multi

array([2.93888937])

In [122]:
# IMPORTANTE = IGNORAMOS el primer coeficiente porque es de la columna de UNOS que agregamos al principio
betas_multi = lr_multi.coef_
betas_multi

array([[ 0.        ,  0.04576465,  0.18853002, -0.00103749]])

In [128]:
beta_1_multi = lr_multi.coef_[0][1]
beta_1_multi

np.float64(0.04576464545539757)

In [129]:
beta_2_multi = lr_multi.coef_[0][2]
beta_2_multi

np.float64(0.18853001691820448)

In [121]:
beta_3_multi = lr_multi.coef_[0][3]
beta_3_multi

np.float64(-0.0010374930424763285)

In [90]:
y_pred_multi = lr_multi.predict(Xcon_unos)

array([[20.52397441],
       [12.33785482],
       [12.30767078],
       [17.59782951],
       [13.18867186],
       [12.47834763],
       [11.72975995],
       [12.12295317],
       [ 3.72734086],
       [12.55084872],
       [ 7.0322992 ],
       [17.28512918],
       [10.57712073],
       [ 8.82630048],
       [18.43436638],
       [20.81929952],
       [12.82365674],
       [23.22495716],
       [ 9.95168206],
       [14.16607293],
       [18.10076728],
       [14.7405382 ],
       [ 6.4891503 ],
       [16.5459329 ],
       [ 8.14651887],
       [15.6100386 ],
       [14.98951429],
       [17.05167344],
       [19.41053803],
       [ 9.14402389],
       [21.6339338 ],
       [11.3460929 ],
       [ 7.63888314],
       [18.86426829],
       [ 7.57483051],
       [17.00682618],
       [23.40590052],
       [15.62347779],
       [ 9.90868103],
       [20.44761039],
       [16.37766467],
       [17.2959832 ],
       [21.59580326],
       [13.96385684],
       [ 8.88787996],
       [15

- Calcular RSS

In [91]:
rss_multi = sum((y - y_pred_multi)**2)
rss_multi

array([556.8252629])

- Calcular RSE


$$\text{RSE} = \sqrt{\frac{1}{n - p}\sum_{i=1}^{n}\left(y_i - \hat{y}_i\right)^2}$$



$$\text{RSE} = \sqrt{\frac{\text{RSS}}{n - p}}$$

$$\text{RSS} = \sum_{i=1}^{n}(y_i - \hat{y}_i)^2$$


In [96]:
rse_multi = np.sqrt(rss_multi / (n_multi - p_multi))
rse_multi

array([1.68551037])

In [98]:
# es una matriz, donde la diagonal equivale a las covarianzas de las beta = la confianza de las beta
# esta matriz realmente no se interpreta
var_beta = np.linalg.inv(Xcon_unos.T @ Xcon_unos) * rse_multi**2
var_beta

array([[ 9.72867479e-02, -2.65727337e-04, -1.11548946e-03,
        -5.91021239e-04],
       [-2.65727337e-04,  1.94573710e-06, -4.47039463e-07,
        -3.26595026e-07],
       [-1.11548946e-03, -4.47039463e-07,  7.41533504e-05,
        -1.78006245e-05],
       [-5.91021239e-04, -3.26595026e-07, -1.78006245e-05,
         3.44687543e-05]])

In [100]:
# errores estándar de los coeficientes = intercepto, TV (variable 1), radio (variable 2) y newspaper (variable 3)
std_beta = np.sqrt(var_beta.diagonal())
std_beta

array([0.31190824, 0.0013949 , 0.00861123, 0.00587101])

Los resultados que se obtuvieron corresponden a los errores estándar de los coeficientes:


$$ \text{SE}(\hat{\beta_0}), \text{SE}(\hat{\beta_1}), \text{SE}(\hat{\beta_2}), \text{SE}(\hat{\beta_3})$$


In [123]:
se_beta_0_multi = std_beta[0]
se_beta_0_multi

np.float64(0.3119082363217914)

In [111]:
se_beta_1_multi = std_beta[1]
se_beta_1_multi

np.float64(0.0013948968069749732)

In [112]:
se_beta_2_multi = std_beta[2]
se_beta_2_multi

np.float64(0.008611233967301957)

In [113]:
se_beta_3_multi = std_beta[3]
se_beta_3_multi

np.float64(0.005871009647086363)

- Calcular estadístico t

In [101]:
# Es igual a 0 porque así se estableció en la hipotésis
mu

0

Para beta 0

In [124]:
t_b0_multi = (beta_0_multi - mu ) / se_beta_0_multi
t_b0_multi

array([9.42228844])

Para beta 1 = TV

In [130]:
t_b1_multi = (beta_1_multi - mu ) / se_beta_1_multi
t_b1_multi

np.float64(32.8086244276697)

Para beta 2 = radio

In [131]:
t_b2_multi = (beta_2_multi - mu ) / se_beta_2_multi
t_b2_multi

np.float64(21.893496058065423)

Para beta 3 = newspaper

In [132]:
t_b3_multi = (beta_3_multi - mu ) / se_beta_3_multi
t_b3_multi

np.float64(-0.1767145865602879)

- Calcular p-value

In [133]:
# tienes beta 0, beta 1, beta 2 y beta 3
factores_multi = 4

In [134]:
p_b0_multi = 2*(1 - stats.t.cdf(np.abs(t_b0_multi), n_multi-factores_multi))
p_b0_multi

array([0.])

In [135]:
p_b1_multi = 2*(1 - stats.t.cdf(np.abs(t_b1_multi), n_multi-factores_multi))
p_b1_multi

np.float64(0.0)

In [136]:
p_b2_multi = 2*(1 - stats.t.cdf(np.abs(t_b2_multi), n_multi-factores_multi))
p_b2_multi

np.float64(0.0)

In [138]:
p_b3_multi = 2*(1 - stats.t.cdf(np.abs(t_b3_multi), n_multi-factores_multi))
p_b3_multi

np.float64(0.8599150500805663)

- Observaciones

Al hacer el cálculo manual de los errores estándar, estadísticos t y p-values bajo el modelo de regresión, se observa que se obtuvieron los mismos resultados que cuando se utilizó la librería statsmodels; por lo tanto se verifica que el procedimiento analítico es consistente con el estimador OLS.